pkg/hub/managedclusterset/rootclustersetcontroller.go

```diff
@@ -0,0 +1,167 @@
+	clusterlisterv1beta1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1beta1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+//go:embed manifests
+var manifestFiles embed.FS
+
+const rootManagedClusterSetName = "RootManagedClusterSet"
+
+type rootManagedClusterSetController struct {
+	kubeClient       kubernetes.Interface
+	clusterClient    clientset.Interface
+	clusterLister    clusterlisterv1.ManagedClusterLister
+	clusterSetLister clusterlisterv1beta1.ManagedClusterSetLister
+	cache            resourceapply.ResourceCache
+	eventRecorder    events.Recorder
+
+	// rootClusterSet caches the mappings between clusters and clustersets.
+	// With the mappings, it's easy to find out which clustersets are impacted on a
+	// change of cluster.
+	rootClusterSet sets.String
+
+	// mapLock protects the read/write of rootClusterSet
+	mapLock sync.RWMutex
+}
+
+func NewRootManagedClusterSetController(
+	kubeClient kubernetes.Interface,
+	clusterClient clientset.Interface,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	clusterSetInformer clusterinformerv1beta1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &rootManagedClusterSetController{
+		kubeClient:       kubeClient,
+		clusterClient:    clusterClient,
+		clusterLister:    clusterInformer.Lister(),
+		clusterSetLister: clusterSetInformer.Lister(),
+		cache:            resourceapply.NewResourceCache(),
+		eventRecorder:    recorder.WithComponentSuffix("root-managed-cluster-set-controller"),
+		rootClusterSet:   sets.NewString(),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterSetInformer.Informer()).
+		WithSync(c.sync).
+		ToController("RootManagedClusterSetController", recorder)
+}
+
+func (c *rootManagedClusterSetController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	rootClusterSetName := rootManagedClusterSetName
+	klog.Infof("Reconciling RootManagedClusterSet")
+
+	rootClusterSet, err := c.clusterSetLister.Get(rootClusterSetName)
+
+	// if rootClusterSet not found, apply it.
+	if errors.IsNotFound(err) {
+		resourceapply.ApplyDirectly(ctx,
+			resourceapply.NewKubeClientHolder(c.kubeClient),
+			syncCtx.Recorder(),
+			c.cache,
+			manifestFiles.ReadFile,
+		)
+	}
+	if err != nil {
+		return err
+	}
+
+	if !rootClusterSet.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	if err := c.syncRootClusterSet(ctx, rootClusterSet); err != nil {
+		return fmt.Errorf("failed to sync RootManagedClusterSet %q: %w", rootClusterSetName, err)
+	}
+	return nil
+}
+
+// syncRootClusterSet syncs root cluster set.
+func (c *rootManagedClusterSetController) syncRootClusterSet(ctx context.Context, originalRootClusterSet *clusterv1beta1.ManagedClusterSet) error {
```

Comment:
In `syncRootClusterSet` function, just compare the spec of the clusterset and see if it has been changed. If it is changed, rollback the change by updating.

You don't have to handle the status of the clusterset. Another controller will do it.